### PR TITLE
Refactor fork selection to be more modular

### PR DESCRIPTION
**Pull request by ParthDesai**
#### Problem
`select_fork` is very long function, incorporating multiple input arguments, and so it is very hard to unit test against. 

#### Summary of Changes
Pull core fork selection logic out of `ReplayStage`, and have it is as its own type, being passed to `ReplayStage` via configuration.

#### Remaining: 
- [x] tests

cc: @aeyakovenko @carllin

### CODE REVIEW

File: core/src/replay_stage.rs [hunk 1]
```diff
@@ -1,5 +1,7 @@
 //! The `replay_stage` replays transactions broadcast by the leader.
 
+use crate::fork_selector::ForkSelector;
+use crate::replay_stage_ds::{ForkProgress, ForkStats};
```
Comment (mvines) [inline core/src/replay_stage.rs]: What does the `_ds` in `replay_stage_ds` mean?   I wonder if we can find a better name

Comment (ParthDesai) [reply core/src/replay_stage.rs]: @mvines It's short for `data structure` :) yeah, I could use a better name.

**Comment by ParthDesai** [comment]
ping @aeyakovenko @carllin

File: core/src/fork_selector.rs [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+    fn extract_forks(
+        &self,
+        frozen_banks: &HashMap<Slot, Arc<Bank>>,
+        ancestors: &HashMap<u64, HashSet<u64>>,
+        descendants: &HashMap<u64, HashSet<u64>>,
+        progress: &HashMap<u64, ForkProgress>,
+    ) -> Vec<Fork> {
+        let mut forks: Vec<Fork> = vec![];
+
+        for (_i, slot) in ancestors.keys().enumerate() {
```
Comment (aeyakovenko) [inline core/src/fork_selector.rs]: where is this logic coming from?

Comment (aeyakovenko) [reply core/src/fork_selector.rs]: its hard to verify that this code does the same thing as the current code.  I was expecting the implementation to match what is currently in replay stage
````
    fn select_fork(                                                
        my_pubkey: &Pubkey,     
        ancestors: &HashMap<u64, HashSet<u64>>,       
        bank_forks: &Arc<RwLock<BankForks>>,                               
        tower: &Tower,                                                   
        progress: &mut HashMap<u64, ForkProgress>,                  
    ) -> VoteAndPoHBank 
```

Comment (aeyakovenko) [reply core/src/fork_selector.rs]: also, you don't need the `enumerate` if you are ignoring the `_i` parameter

File: core/src/fork_selector.rs [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+            if (!descendants.contains_key(slot) || descendants[slot].is_empty())
+                && frozen_banks.contains_key(slot)
```
Comment (aeyakovenko) [inline core/src/fork_selector.rs]: this check can be on the bank, is_frozen

File: core/src/replay_stage.rs [hunk 4]
```diff
@@ -792,34 +717,64 @@ impl ReplayStage {
...
+        frozen_banks: &HashMap<Slot, Arc<Bank>>,
+        progress: &HashMap<u64, ForkProgress>,
+    ) -> VoteAndPoHBank
+    where
+        T: ForkSelector,
+    {
+        let forks = fork_selector.extract_forks(
+            frozen_banks,
+            ancestors,
+            &bank_forks.read().unwrap().descendants(),
```
Comment (aeyakovenko) [inline core/src/replay_stage.rs]: decedents where previously not necessary to do selection.

File: core/src/replay_stage_ds.rs [hunk 5]
```diff
@@ -0,0 +1,98 @@
+use crate::consensus::StakeLockout;
```
Comment (aeyakovenko) [inline core/src/replay_stage_ds.rs]: What is replay_stage_ds?  typically we name modules after the top most class

File: core/src/fork_selector.rs [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+    /// finds best fork out of vector of fork
+    fn find_best_fork(
+        &self,
+        forks: Vec<Fork>,
+        progress: &HashMap<u64, ForkProgress>,
+    ) -> Option<Fork>;
+}
+
+#[derive(Default)]
+pub struct OptimalForkSelector {}
```
Comment (aeyakovenko) [inline core/src/fork_selector.rs]: This name should be descriptive of the algorithm, not that its the "best" or "optimal"

File: core/src/fork_selector.rs (hunk 2 above)
Comment (aeyakovenko) [reply core/src/fork_selector.rs]: ah I see, the select_fork function is renamed to compute_fork_progress

Comment (aeyakovenko) [inline core/src/fork_selector.rs]: This loop is weird.  It makes more sense to me for 
* compute_progress to update the ForkStats data structure
* select_fork to iterate over frozen_banks and select the best fork based on the ForkStats.

File: core/src/fork_selector.rs [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+        &self,
+        frozen_banks: &HashMap<Slot, Arc<Bank>>,
+        ancestors: &HashMap<u64, HashSet<u64>>,
+        descendants: &HashMap<u64, HashSet<u64>>,
+        progress: &HashMap<u64, ForkProgress>,
+    ) -> Vec<Fork> {
+        let mut forks: Vec<Fork> = vec![];
+
+        for (_i, slot) in ancestors.keys().enumerate() {
+            if (!descendants.contains_key(slot) || descendants[slot].is_empty())
```
Comment (aeyakovenko) [inline core/src/fork_selector.rs]: do you need to do these checks after ForkStats has been computed?

**Pull request closed without merging**